app/src/main/java/com/ddoczi/tasky/agenda/presentation/detail/agenda/AgendaDetailViewModel.kt

```diff
@@ -8,88 +9,177 @@ import com.ddoczi.tasky.agenda.domain.repository.ReminderRepository
+                            AgendaType.REMINDER -> reminderRepository.insertReminder(state.value.agendaItem as AgendaItem.Reminder)
+                            AgendaType.EVENT -> eventRepository.insertEvent(state.value.agendaItem as AgendaItem.Event)
+                        }
+                        _state.update { it.copy(shouldExit = true) }
                     }
-                    _state.update { it.copy(shouldExit = true) }
+                } catch (e: Exception) {
+                    println(e)
                 }
             }
             is AgendaDetailEvent.OnDelete -> {
-                viewModelScope.launch(NonCancellable) {
-                    when (state.value.agendaType) {
-                        AgendaType.TASK -> taskRepository.deleteTaskById(state.value.id)
-                        AgendaType.REMINDER -> reminderRepository.deleteReminderById(state.value.id)
-                        AgendaType.EVENT -> eventRepository.deleteEventById(state.value.id)
+                try {
+                    viewModelScope.launch(NonCancellable) {
+                        if (state.value.agendaItem?.agendaItemId?.isEmpty() == true) return@launch
+                        when (state.value.agendaType) {
+                            AgendaType.TASK -> state.value.agendaItem?.let {
+                                taskRepository.deleteTaskById(
+                                    it.agendaItemId)
+                            }
+                            AgendaType.REMINDER -> state.value.agendaItem?.let {
+                                reminderRepository.deleteReminderById(
+                                    it.agendaItemId)
+                            }
+                            AgendaType.EVENT -> state.value.agendaItem?.let {
+                                eventRepository.deleteEventById(
+                                    it.agendaItemId)
+                            }
+                        }
+                        _state.update { it.copy(shouldExit = true) }
                     }
-                    _state.update { it.copy(shouldExit = true) }
+                } catch (e: Exception) {
+                    println(e)
                 }
             }
             is AgendaDetailEvent.OnFromDateSelected -> {
-                _state.update { it.copy(fromDate = event.fromDate) }
+                _state.update {
+                    it.copy(
+                        fromDate = event.fromDate,
+                        agendaItem = when(state.value.agendaType) {
+                            AgendaType.EVENT -> itemAsEvent()?.copy(eventFromDate = event.fromDate.atTime(state.value.time))
+                            AgendaType.REMINDER -> itemAsReminder()?.copy(reminderDate = event.fromDate.atTime(state.value.time))
+                            AgendaType.TASK -> itemAsTask()?.copy(taskDate = event.fromDate.atTime(state.value.time))
+                        }
```

Comment:
Alternatively, you could think about creating a sealed hierarchy just for the differences of the agenda items. This is what I mean:
```
sealed interface AgendaItemDetails {
    data class Event(
        val toTime: LocalDateTime,
        val attendees: List<Attendee>,
        val photos: List<Photo>
    ): AgendaItemDetails
    
    data class Task(
        val isDone: Boolean
    ): AgendaItemDetails
    
    data object Reminder: AgendaItemDetails
}
```
That way, you just need type-checks when updating properties that are individual to one of the agenda item types, but you don't need this for updating shared properties anymore like the title:
```
data class AgendaDetailState(
    val title: String = "",
    val description: String = "",
    val time: LocalDateTime = LocalDateTime.now(),
    val details: AgendaItemDetails? = null
)
```
Updating then works like this:
```
// Updating a shared property
state = state.copy(title = "new title")

// Updating an individual property
state = state.copy(
    details = detailsAsEvent()?.copy(attendees = newAttendees)
)
```